src/flint/types/fq_default_poly.pyx

```diff
@@ -0,0 +1,1655 @@
+
+            >>> R = fq_default_poly_ctx(163)
+            >>> x = R.gen()
+            >>> f = (x - 1) * (x - 2)**3 * (x - 3)**5
+            >>> f.roots()
+            [(1, 1), (2, 3), (3, 5)]
+            >>> f.roots(multiplicities=False)
+            [1, 2, 3]
+        """
+        cdef fq_default_poly_factor_t fac
+        cdef int i, with_multiplicity
+
+        with_multiplicity = 0
+        if multiplicities:
+            with_multiplicity = 1
+
+        fq_default_poly_factor_init(fac, self.ctx.field.val)
+        fq_default_poly_roots(fac, self.val, with_multiplicity, self.ctx.field.val)
+
+        num = fq_default_poly_factor_length(fac, self.ctx.field.val)
+        res = [0] * num
+
+        cdef fq_default_poly linear_factor
+        cdef fq_default root
+
+        for i in range(num):
+            # Get a factor of the form (x - a)
+            linear_factor = self.ctx.new_ctype_poly()
+            fq_default_poly_factor_get_poly(linear_factor.val, fac, i, self.ctx.field.val)
+
+            # Compute a from (x - a)
+            root = self.ctx.field.new_ctype_fq_default()
+            fq_default_poly_get_coeff(
+                root.val, linear_factor.val, 0, root.ctx.val
+            )
+            fq_default_neg(
+                root.val, root.val, root.ctx.val
+            )
+            if multiplicities:
+                mul = fq_default_poly_factor_exp(fac, i, self.ctx.field.val)
+                res[i] = (root, mul)
+            else:
+                res[i] = root
+        return res
+
+    def complex_roots(self):
+        """
+        This method is not implemented for polynomials in finite fields
+        """
+        raise DomainError("Cannot compute complex roots for polynomials over finite fields")
+
+    # ====================================
+    # Inflation and Deflation
+    # ====================================
+
+    def inflate(self, ulong n):
+        r"""
+        Returns the result of the polynomial `f = \textrm{self}` to
+        `f(x^n)`
+
+            >>> R = fq_default_poly_ctx(163, 3)
+            >>> f = R([1,2,3])
+            >>> f.inflate(10)
+            3*x^20 + 2*x^10 + 1
+
+        """
+        cdef fq_default_poly res
+        res = self.ctx.new_ctype_poly()
+        fq_default_poly_inflate(
+            res.val, self.val, n, res.ctx.field.val
+        )
+        return res
+
+    def deflate(self, ulong n):
+        r"""
+        Returns the result of the polynomial `f = \textrm{self}` to
+        `f(x^{1/n})`
+
+            >>> R = fq_default_poly_ctx(163, 3)
+            >>> f = R([1,0,2,0,3])
+            >>> f
+            3*x^4 + 2*x^2 + 1
+            >>> f.deflate(2)
+            3*x^2 + 2*x + 1
+
+        """
+        cdef fq_default_poly res
+
+        n_max = fq_default_poly_deflation(
+            self.val, self.ctx.field.val
+        )
+        if n > n_max:
+            raise ValueError(f"Cannot deflate with {n = }, maximum allowed value is {n_max = }")
```

Comment:
Is this check necessary to prevent a core dump or something?

Reply:
I think so, I had to do it for fmpz_mod and I just assumed the same check would need to be here (It will definitely need to be for the FMPZ_MOD type I suppose if nothing else...)